backends/dpdk/dpdkProgram.cpp

```diff
@@ -240,41 +313,59 @@ bool ConvertToDpdkParser::preorder(const IR::P4Parser *p) {
-                                             left->to<IR::Constant>()->asUnsigned();
-                            auto tmpDecl = addNewTmpVarToMetadata("tmpMask", switch_var->type);
-                            auto tmpMask = new IR::PathExpression(
-                                               IR::ID("m." + tmpDecl->name.name));
-                            collector->push_variable(new IR::DpdkDeclaration(tmpDecl));
-                            add_instr(new IR::DpdkMovStatement(tmpMask, switch_var));
-                            add_instr(new IR::DpdkAndStatement(tmpMask, tmpMask, right));
+                const IR::Expression *caseExpr;
+                IR::Expression *left;
+                IR::Expression *right;
+
+                /* Handling single expression in keyset as special case because :
+                   - for tuple expression we emit a series of jmpneq followed by an unconditional
+                     jump, handling single expression separately by emitting a jmpeq saves us one
+                     unconditional jmp
+                   - the keyset for single expression is not a ListExpression, so handling it with
+                     tuple expressions would anyways require additional and conversion to
+                     ListExpression.
+                */
+                if (e->select->components.size() == 1) {
+                    switch_var = e->select->components.at(0);
+                    for (auto sc : caseList) {
+                        caseExpr = sc->keyset;
+                        if (!sc->keyset->is<IR::DefaultExpression>()){
+                            getCondVars(switch_var, caseExpr, &left, &right);
                             add_instr(new IR::DpdkJmpEqualStatement(
-                                         append_parser_name(p, v->state->path->name),
-                                                            tmpMask, new IR::Constant(value)));
+                                      append_parser_name(p, sc->state->path->name), left, right));
                         } else {
-                            add_instr(new IR::DpdkJmpEqualStatement(
-                              append_parser_name(p, v->state->path->name), switch_var, v->keyset));
+                            add_instr(new IR::DpdkJmpLabelStatement(
+                                      append_parser_name(p, sc->state->path->name)));
+                        }
+                    }
+                } else {
+                    auto tupleInputExpr = e->select;
+                    auto inputSize = tupleInputExpr->components.size();
+                    auto count = 0;
+                    cstring trueLabel, falseLabel;
+                    /* For each select case, emit the block start label and then a series of
+                       jmp instructions. */
+                    for (auto sc : caseList) {
+                        cstring labelName = state->name + "_" + std::to_string(count++);
```

Comment:
This looks like it could generate collisions if some states already have such names. I would always recommend using a name generator, but you have to use it everywhere so it keeps track of all used symbols.

Reply:
I have updated the code to use refmap->newName() for generating the case label.